src/setup_payload/AdditionalDataPayloadGenerator.cpp

```diff
@@ -0,0 +1,137 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file provides a utility to generate Additional Data payload and its members
+ *      (e.g. rotating device id)
+ *
+ */
+
+#include "AdditionalDataPayloadGenerator.h"
+#include "AdditionalDataPayload.h"
+
+#include <core/CHIPCore.h>
+#include <core/CHIPTLV.h>
+#include <crypto/CHIPCryptoPAL.h>
+#include <stdlib.h>
+#include <support/BufferWriter.h>
+#include <support/CHIPMem.h>
+
+using namespace chip;
+using namespace chip::System;
+using namespace chip::TLV;
+using namespace chip::Crypto;
+using namespace chip::SetupPayload;
+using namespace chip::Encoding::LittleEndian;
+
+static uint8_t * sHashInput;
+
+AdditionalDataPayloadGenerator::AdditionalDataPayloadGenerator(uint16_t lifetimeCounter, char * serialNumberBuffer,
+                                                               size_t serialNumberBufferSize) :
+    mLifetimeCounter(lifetimeCounter),
+    mSerialNumberBuffer(serialNumberBuffer), mSerialNumberBufferSize(serialNumberBufferSize)
+{
+    sHashInput =
```

Comment:
This is still dynamic allocation. There should be a fixed size buffer that can accomodate the maximal serial number.

Suggest avoiding the reference to `serialNumberBuffer` in constructor, and just pass all values as more arguments to `generateAdditionalDataPayload`, so that it could even all be omitted from storage and remain on the stack at the call site.

Reply:
updated